### PR TITLE
Build URL params and include requestId

### PR DESCRIPTION
**Pull request by gtk-grafana**
In order to facilitate performance observer tracking of dashboard queries we need unique identifiers for each dashboard request. This PR facilities allowing datasources to provide a requestId which will be appended to the query request as a URL parameter.

### CODE REVIEW

File: packages/grafana-runtime/src/utils/DataSourceWithBackend.ts
Comment (ryantxu) [inline packages/grafana-runtime/src/utils/DataSourceWithBackend.ts:207]:
```suggestion
    let url = '/api/ds/query?type='+this.id;
```
then the rest can be:
```
url += '&key=var'
```

and type is the one i am most curious to see in the network tab!

Comment (leeoniya) [reply packages/grafana-runtime/src/utils/DataSourceWithBackend.ts]: `type` is a bit ambiguous here. is it query type or ds type? maybe dstype?

doesnt matter if query is always same type, but could be instant vs range?

Comment (gtk-grafana) [reply packages/grafana-runtime/src/utils/DataSourceWithBackend.ts]: now `ds_type`

File: packages/grafana-runtime/src/utils/DataSourceWithBackend.ts
Comment (ryantxu) [inline packages/grafana-runtime/src/utils/DataSourceWithBackend.ts:211]:
```suggestion
      url += '&expression=true';
```

Comment (ryantxu) [reply packages/grafana-runtime/src/utils/DataSourceWithBackend.ts:211]: then get rid of URLSearchParams... we don't have anything with crazy escaping etc

Comment (gtk-grafana) [reply packages/grafana-runtime/src/utils/DataSourceWithBackend.ts:211]: Uh why tho? it's already bundled up in the queryStr

Comment (gtk-grafana) [reply packages/grafana-runtime/src/utils/DataSourceWithBackend.ts:211]: I guess we could avoid using the `URLSearchParams` altogether since we've always got a ds_type

Comment (gtk-grafana) [reply packages/grafana-runtime/src/utils/DataSourceWithBackend.ts]: Sorry, github didn't serve me your second commend right away. Fixed

**Comment by gtk-grafana** [comment]
Aiming for 9.5.1?

File: packages/grafana-runtime/src/utils/DataSourceWithBackend.ts
Comment (ryantxu) [inline packages/grafana-runtime/src/utils/DataSourceWithBackend.ts:205]: unused 🤔   but otherwise LGTM

**Comment by ryantxu** [comment]
LGTM, but lets wait till after monday to merge

File: packages/grafana-runtime/src/utils/DataSourceWithBackend.ts
Comment (andresmgot) [inline packages/grafana-runtime/src/utils/DataSourceWithBackend.ts:214]: as far as I can tell, data sources cannot access URL params set in the query and that's why "from expression" was added both as a header and a URL param in https://github.com/grafana/grafana/pull/65306. Why not using headers here to avoid ending in the same situation?

Comment (gtk-grafana) [reply packages/grafana-runtime/src/utils/DataSourceWithBackend.ts:214]: I believe this is because in the `PerformanceObserver` all we have access to is the url, and we don't have access to headers in that context. @leeoniya is that correct?

Comment (leeoniya) [reply packages/grafana-runtime/src/utils/DataSourceWithBackend.ts:214]: yes, this is purely for frontend tracking. these requestIds are not unique, they simply increment from initial dashboard load (Q100, Q101, Q102...). when we construct these requests we need to somehow correlate them from an async PerformanceObserver callback so we can measure stuff (like transfer savings). unfortunately PerformanceObserver entries don't give us access to anything except the query URL, so this is the only place we can stick it to make that link. you can see what we do with it in https://github.com/grafana/grafana/pull/65244

these ids aren't useful in the backend, but if they ever become UIDs and useful, we can always add the header as well in the future.

Comment (gtk-grafana) [reply packages/grafana-runtime/src/utils/DataSourceWithBackend.ts:214]: @ryantxu ^^ This is the direct opposite of your request to make this available for all data sources.

Comment (leeoniya) [reply packages/grafana-runtime/src/utils/DataSourceWithBackend.ts:214]: @gtk-grafana already has an Influx PR that will use this also, next will be loki/logs. after that it can be any datasource that returns timestamps in the data.

in general, this strategy allows us to cheaply collect perf metrics of any DS request without delegating this to the backend which in most cases is infeasible since they stream responses and cannot measure end-to-end performance as the user experiences it.

Comment (andresmgot) [reply packages/grafana-runtime/src/utils/DataSourceWithBackend.ts:214]: I don't mean to block this, if it will be used widely please go ahead. Also, please evaluate the suggestion at https://github.com/grafana/grafana/pull/65244#issuecomment-1496161971 in case it's helpful.

**Review by ryantxu — APPROVED** [review]
LGTM -- @andresmgot , I get your point about this feeling prometheus specific.  I struggle with that a bit because that is the thing we want to measure today, but need this kind of thing in other cases.  Either we pollute the public API with "but I want X for this case" or come up with some patterns that just exist.  I am in favor of the 2nd :)

**Comment by gtk-grafana** [comment]
Merging in targeted at g10, so hopefully if this causes any issues we'll find out before then.

**Pull request merged**
